v2/pkg/engine/resolve/resolve.go

```diff
@@ -412,11 +420,58 @@ func (r *Resolver) handleEvent(event subscriptionEvent) {
 		r.handleTriggerInitialized(event.triggerID)
 	case subscriptionEventKindTriggerShutdown:
 		r.handleTriggerShutdown(event)
+	case subscriptionEventKindHeartbeat:
+		r.handleHeartbeat(event.triggerID, event.data)
 	case subscriptionEventKindUnknown:
 		panic("unknown event")
 	}
 }
 
+func (r *Resolver) handleHeartbeat(id uint64, data []byte) {
+	trig, ok := r.triggers[id]
+	if !ok {
+		return
+	}
+	if r.options.Debug {
+		fmt.Printf("resolver:heartbeat:%d\n", id)
+	}
+	wg := &sync.WaitGroup{}
+	trig.inFlight = wg
+	for c, s := range trig.subscriptions {
+		c, s := c, s
+		if err := r.triggerUpdateSem.Acquire(r.ctx, 1); err != nil {
+			return
+		}
+		wg.Add(1)
+		go func() {
+			defer r.triggerUpdateSem.Release(1)
+			defer wg.Done()
+
+			if r.options.Debug {
+				fmt.Printf("resolver:heartbeat:subscription:%d\n", s.id.SubscriptionID)
+			}
+
+			if _, err := s.writer.Write(data); err != nil {
+				r.asyncErrorWriter.WriteError(c, err, nil, s.writer)
+			}
+			s.mux.Lock()
+			err := s.writer.Flush()
+			s.mux.Unlock()
+			if err != nil {
+				// client disconnected
+				_ = r.AsyncUnsubscribeSubscription(s.id)
+				return
+			}
+			if r.options.Debug {
+				fmt.Printf("resolver:heartbeat:subscription:flushed:%d\n", s.id.SubscriptionID)
+			}
+			if r.reporter != nil {
+				r.reporter.SubscriptionUpdateSent()
```

Comment:
Should I define `SubscriptionHeartbeatSent`? Would that be a valuable metric?

Reply:
I'd say no, at least I couldn't make use of it right now.